feat: add initial metric publisher

wulfmann opened the pull request:
*Issue #, if available:*

#22 

*Description of changes:*

Adds an initial metrics publisher

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

wulfmann commented:
Let me know if you'd like to go a different direction with this. It is in line with the java and go plugins.

rjlohan commented:
Thanks for this contribution! Just a few test errors in the build. You can reproduce the test failures locally by running;

```
pytest  
    --cov-report=term \ 
    --cov-report=html \ 
    --cov="rpdk.python" \ 
    --cov="cloudformation_cli_python_lib" \
    "tests/"
```

Once those are passing we can look at merging this in.

wulfmann commented:
Yea I forgot to install the precommit. I’m working through the errors now and I’ll circle back when I get them fixed.

Thanks!

wulfmann commented:
@rjlohan i've pushed updates here and the tests should pass now. I updated the .gitignore as it was disallowing new tests being added, and i added a test for the metric publisher.

Let me know if you like this approach for the publisher. I haven't added in any of the calls to actual handler yet, this is just the class. If we like this direction then i can add the calls into the actual handler.

tobywf commented:
Hey, thanks for the PR, it's nice and clean! We're working pretty hard getting some more stuff ready and wanted to get the Python stuff out there to let people know we're working on it and it's a priority. Apologies if it takes me a bit to get to this!

wulfmann commented:
@tobywf take your time no worries!

jaymccon commented:
looks good 😄 Is the intent to wire this up to `Resource.__call__` in a future pr ?

wulfmann commented:
@jaymccon that was my thinking yes. I originally had it on this pr, but removed it as it changed a lot of function signatures in the other tests. I think it'll be a quicker pr in that case.

jaymccon commented on src/cloudformation_cli_python_lib/metrics.py:
```diff
@@ -0,0 +1,96 @@
+LOG = logging.getLogger(__name__)
+
+
+def format_dimensions(dimensions: Mapping[Any, Any]) -> Any:  # TODO fix type
```
from https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Client.put_metric_data it looks like key and value must both be strings, so `Mapping[Str, Str] -> List[Mapping[Str, Str]]` should do the trick here

wulfmann replied:
updated

tobywf commented on src/cloudformation_cli_python_lib/metrics.py:
```diff
@@ -0,0 +1,96 @@
+    formatted_dimensions = []
+    for key, value in dimensions.items():
+        formatted_dimensions.append({"Name": key, "Value": value})
+    return formatted_dimensions
```
would love this to be a list comprehension

wulfmann replied:
fixed

tobywf commented on src/cloudformation_cli_python_lib/metrics.py:
```diff
@@ -0,0 +1,96 @@
+                    }
+                ],
+            )
+            LOG.debug(res)
```
does this help resource authors? bit worried about log spam

wulfmann replied:
put_metric_data actually returns None so i've removed the log.

tobywf commented on src/cloudformation_cli_python_lib/metrics.py:
```diff
@@ -0,0 +1,96 @@
+    def publish_exception_metric(self, date: float, action: Action, error: Any) -> None:
+        dimensions = {
+            "DimensionKeyActionType": action,
+            "DimensionKeyExceptionType": str(error),
```
this isn't quite the same as the Java plugin (`e.getClass().toString()`). if `error: Exception` (or maybe even `BaseException`, since the Java type is Throwable), a closer analogue would be `error.__class__` or `type(error)` i think?

wulfmann replied:
i went with str(type(error)). You're right it wasn't exactly like the java implementation.

tobywf commented on tests/lib/metrics_test.py:
```diff
@@ -0,0 +1,130 @@
+    def __init__(self, client):
+        self._client = client
+
+    def client(self, name):  # pylint: disable-msg=unused-argument
```
you can use `_name` here to mark arguments as unused. pylint will respect that convention

wulfmann replied:
Good call, updated.

tobywf commented on src/cloudformation_cli_python_lib/metrics.py:
```diff
@@ -0,0 +1,96 @@
+        dimensions: Mapping[Any, Any],  # TODO: fix type
+        unit: StandardUnit,
+        value: Any,
+        date: float,
```
might be good to call this `timestamp`? i think the type is `datetime.datetime`, [at least the boto3 docs have an example with that](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Client.put_metric_data). maybe float works, too? the documentation isn't exactly obvious:

> Timestamp (datetime) -- The time the metric data was received, expressed as the number of milliseconds since Jan 1, 1970 00:00:00 UTC.

also, `value` seems like it should be `float`, not `Any`

wulfmann replied:
updated

wulfmann commented:
I could use some feedback on the first test that should make sure that an exception during the put_metric_data call is logged. I attempted to mock the LOG variable for awhile and don’t have a ton of luck. Do you see anything obvious that would make that easy to check? Currently the test does nothing. That’s the last main TODO here.

jaymccon commented:
> I attempted to mock the LOG variable for awhile and don’t have a ton of luck. 

patching `LOG` should work, something like:

```python
@patch("cloudformation_cli_python_lib.metrics.LOG", autospec=True)
def test_put_metric_catches_error(mock_log):
    ...
    mock_log.error.assert_called_once()
```

wulfmann commented:
I had tried that and don't know what I was doing wrong, but it works now. This PR should be up to date now with requested changes.

jaymccon commented on tests/lib/metrics_test.py:
```diff
@@ -0,0 +1,144 @@
+    expected_calls = [
+        call.error(
+            "An error occurred while publishing metrics: %s",
+            "An error occurred (InternalServiceError) when calling the PutMetricData operation: ",  # noqa: B950 # pylint: disable=line-too-long
```
can split that string into 2 lines to avoid the lint excludes

```
            "An error occurred while publishing metrics: %s",
            "An error occurred (InternalServiceError) when calling the "
            "PutMetricData operation: ",
```

wulfmann replied:
nice tip, updated
